porch/func/internal/podevaluator.go

```diff
@@ -65,30 +74,109 @@ func NewPodEvaluator(namespace, wrapperServerImage string) (Evaluator, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to create client: %w", err)
 	}
-	return &podEvaluator{
+
+	pe := &podEvaluator{
 		kubeClient:         cl,
 		namespace:          namespace,
 		wrapperServerImage: wrapperServerImage,
-	}, nil
+		gcScanInternal:     time.Duration(interval) * time.Second,
+		podTTL:             time.Duration(ttl) * time.Second,
+		imageEntrypoint:    map[string][]string{},
+		grpcConnectionPool: map[string]*grpc.ClientConn{},
+	}
+	go pe.garbageCollector()
+	return pe, nil
 }
 
-func (pe *podEvaluator) EvaluateFunction(ctx context.Context, req *pb.EvaluateFunctionRequest) (*pb.EvaluateFunctionResponse, error) {
-	ref, err := name.ParseReference(req.Image)
-	if err != nil {
-		return nil, err
+// TODO: We can use Watch + periodically reconciliation to manage the pods,
+// the pod evaluator will become a controller.
+func (pe *podEvaluator) garbageCollector() {
+	var err error
+	podList := &corev1.PodList{}
+	for {
+		time.Sleep(pe.gcScanInternal)
+		err = pe.kubeClient.List(context.Background(), podList, client.InNamespace(pe.namespace))
+		if err != nil {
+			klog.Warningf("unable to list pods in namespace %v: %w", pe.namespace, err)
+			continue
+		}
+		for _, pod := range podList.Items {
+			lastUse, found := pod.Annotations[lastUseTimeAnnotation]
+			// If a pod doesn't have a last-use annotation, we patch it.
+			if !found {
+				if err = patchPodWithUnixTimeAnnotation(context.Background(), pe.kubeClient, &pod); err != nil {
+					klog.Warningf("can't patch pod %v/%v with the last-use annotation: %w", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				lu, err := strconv.ParseInt(lastUse, 10, 64)
+				if err != nil {
+					klog.Warningf("unable to convert the Unix time string to int64: %w", err)
+					// If the annotation is ill-formatted, we patch it with the current time and will try to GC it later.
+					if err = patchPodWithUnixTimeAnnotation(context.Background(), pe.kubeClient, &pod); err != nil {
+						klog.Warningf("can't patch pod %v/%v with the last-use annotation: %w", pod.Namespace, pod.Name, err)
+					}
+					continue
+				}
+				if time.Unix(lu, 0).Add(pe.podTTL).Before(time.Now()) {
+					// deleting the grpc connection from the pool
+					image := pod.Spec.Containers[0].Image
+					cc, found := pe.grpcConnectionPool[image]
+					if found {
+						klog.Infof("deleting grpc connection %v for %v", cc.Target(), image)
+						delete(pe.grpcConnectionPool, image)
```

Comment:
we'll need synchronization between the garbageCollector and the request handler goroutines

this will need to handle cases not only of concurrent access and updates, but also concurrent cache misses with subsequent pod spin-up.

Ideally we need logic such as:

```
gRPCClient, found := findCachedClientForFunction(image)
if !found {
  gRPCClient = coordinatePodSpinUpAndWakeMeWhenClientAvailable(image)
}
```

the coordinated spin-up should block all but one request as they wait for the pod to come up, while the running goroutine does the spin-up.

One way to do it is to have a goroutine dedicated to managing the image cache and goroutines handling the requests get the gRPC client(s) via channels. That way the synchronization may be easiest.

Reply:
> we'll need synchronization between the garbageCollector and the request handler goroutines

Yes, you are right!

Re. coordinated spin-up, there are 2 levels of coordination we may need to spin up pods:
- Coordination between goroutines In one function runner
- Coordination between function runners

To achieve the later coordination, we may want to use k8s's coordination API: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#lease-v1-coordination-k8s-io